icechunk/src/ops/gc.rs

```diff
@@ -153,7 +185,141 @@ pub enum GCError {
+        Arc::clone(&asset_manager),
+        &config.extra_roots,
+    )
+    .await?;
+
+    let keep_chunks = Arc::new(Mutex::new(HashSet::new()));
+    let keep_manifests = Arc::new(Mutex::new(HashSet::new()));
+    let keep_snapshots = Arc::new(Mutex::new(HashSet::new()));
+
+    let all_manifest_infos = all_snaps
+        .map(ready)
+        .buffer_unordered(config.max_snapshots_in_memory.get() as usize)
+        // this could be slightly optimized by not collecting all manifest info records into a vec
+        // but we don't expect too many, and they are small anyway
+        .and_then(|snap| {
+            let keep_snapshots = Arc::clone(&keep_snapshots);
+            async move {
+                {
+                    keep_snapshots
+                        .lock()
+                        .map_err(|_| {
+                            RepositoryError::from(RepositoryErrorKind::Other(
+                                "can't lock retained snapshots mutex".to_string(),
+                            ))
+                        })?
+                        .insert(snap.id());
+                }
+                Ok(stream::iter(
+                    snap.manifest_files()
+                        .map(Ok::<_, RepositoryError>)
+                        .collect::<Vec<_>>(),
+                ))
+            }
+        })
+        .try_flatten();
+
+    let manifest_infos = try_unique_stream(|mi| mi.id.clone(), all_manifest_infos);
+
+    // we want to fetch many manifests in parallel, but not more than memory allows
+    // for this we use the StreamLimiter using the manifest size in bytes for usage
+    let limiter = &Arc::new(StreamLimiter::new(
+        "garbage_collect".to_string(),
+        config.max_compressed_manifest_mem_bytes.get(),
+    ));
+
+    let keep_man = &keep_manifests;
+    let asset_man = &asset_manager;
+    let keep_chunks_c = Arc::clone(&keep_chunks);
+    let compute_stream = limiter
+        .limit_stream(manifest_infos, |minfo| minfo.size_bytes as usize)
+        .map_ok(|m| async move {
+            {
+                keep_man
+                    .lock()
+                    .map_err(|_| {
+                        RepositoryError::from(RepositoryErrorKind::Other(
+                            "can't lock retained manifests mutex".to_string(),
+                        ))
+                    })?
+                    .insert(m.id.clone());
+            }
+            let manifest =
+                Arc::clone(asset_man).fetch_manifest(&m.id, m.size_bytes).await?;
+            Ok((manifest, m))
+        })
+        // Now we can buffer a bunch of fetch_manifest operations. Because we are using
+        // StreamLimiter we know memory is not going to blow up
+        .try_buffer_unordered(config.max_concurrent_manifest_fetches.get() as usize)
+        .and_then(move |(manifest, minfo)| {
+            let keep_chunks_c = Arc::clone(&keep_chunks_c);
+            async move {
+                task::spawn_blocking(move || {
```

Comment:
Is this necessary because there could be millions of chunks and iter would block otherwise?

Reply:
yeah ... a manifest could have 10M chunks or something, this would take a while and in the meantime some HTTP requests in flight could timeout